src/compiler/factory/emitHelpers.ts

```diff
@@ -1367,6 +1387,66 @@ export const classPrivateFieldInHelper: UnscopedEmitHelper = {
             };`
 };
 
+/**
+ * @internal
+ */
+export const addDisposableResourceHelper: UnscopedEmitHelper = {
+    name: "typescript:addDisposableResource",
+    importName: "__addDisposableResource",
+    scoped: false,
+    text: `
+        var __addDisposableResource = (this && this.__addDisposableResource) || function (env, value, async) {
+            if (value !== null && value !== void 0) {
+                if (typeof value !== "object") throw new TypeError("Object expected.");
+                var dispose;
+                if (async) {
+                    if (!Symbol.asyncDispose) throw new TypeError("Symbol.asyncDispose is not defined.");
+                    dispose = value[Symbol.asyncDispose];
+                }
+                if (dispose === void 0) {
+                    if (!Symbol.dispose) throw new TypeError("Symbol.dispose is not defined.");
+                    dispose = value[Symbol.dispose];
+                }
+                if (typeof dispose !== "function") throw new TypeError("Object not disposable.");
+                env.stack.push({ value: value, dispose: dispose, async: async });
+            }
+            else if (async) {
+                env.stack.push({ async: true });
+            }
+            return value;
+        };`
+};
+
+/**
+ * @internal
+ */
+export const disposeResourcesHelper: UnscopedEmitHelper = {
+    name: "typescript:disposeResources",
+    importName: "__disposeResources",
+    scoped: false,
+    text: `
+        var __disposeResources = (this && this.__disposeResources) || function (env) {
+            function fail(e) {
+                env.error = env.hasError ? new SuppressedError(e, env.error) : e;
+                env.hasError = true;
+            }
+            function next() {
+                while (env.stack.length) {

```

Comment:
I could copy your idea to optimize away the `next()` tail recursion that Babel uses 👀